### ISSUE TITLE
1.14: heap use after free: start tutorial, quit to main menu, start tutorial

### ISSUE THREAD

**Issue by matthiaskrgr** [wesnoth/wesnoth#2903]
game @ 7821f470c2c70eb0865f27b3baece9d0ce91afbc
start game in small resolution
start tutorial
=> get warning that window is too small
````
../src/gui/widgets/window.cpp:762 in function 'layout' found the following problem: Failed to size window; wanted size 1255,700 available size 956,1041 screen size 956,1041.
````
dismiss warning
try to restart tutorial with same window size
=> boom
````
=================================================================
==17470==ERROR: AddressSanitizer: heap-use-after-free on address 0x6080009a37f8 at pc 0x0000054235b6 bp 0x7ffee7065f60 sp 0x7ffee7065f58
READ of size 4 at 0x6080009a37f8 thread T0
    #0 0x54235b5 in surface::add_surface_ref(SDL_Surface*) /home/matthias/vcs/github/wesnoth/build/../src/sdl/surface.hpp:85:4
    #1 0x5423445 in surface::assign_surface_internal(SDL_Surface*) /home/matthias/vcs/github/wesnoth/build/../src/sdl/surface.hpp:91:3
    #2 0x5418ef0 in surface::assign(surface const&) /home/matthias/vcs/github/wesnoth/build/../src/sdl/surface.hpp:53:3
    #3 0x666cec9 in surface::operator=(surface const&) /home/matthias/vcs/github/wesnoth/build/../src/sdl/surface.hpp:58:3
    #4 0x931ec48 in image::get_image(image::locator const&, image::TYPE) /home/matthias/vcs/github/wesnoth/build/../src/image.cpp:1037:10
    #5 0x98adcc7 in gui2::image::calculate_best_size() const /home/matthias/vcs/github/wesnoth/build/../src/gui/widgets/image.cpp:46:16
    #6 0x72933f4 in gui2::widget::get_best_size() const /home/matthias/vcs/github/wesnoth/build/../src/gui/widgets/widget.cpp:195:12
    #7 0x70dffe4 in gui2::grid::child::get_best_size() const /home/matthias/vcs/github/wesnoth/build/../src/gui/widgets/grid.cpp:742:35
    #8 0x70db55f in gui2::grid::calculate_best_size() const /home/matthias/vcs/github/wesnoth/build/../src/gui/widgets/grid.cpp:436:43
    #9 0x72933f4 in gui2::widget::get_best_size() const /home/matthias/vcs/github/wesnoth/build/../src/gui/widgets/widget.cpp:195:12
    #10 0x703af72 in gui2::container_base::calculate_best_size() const /home/matthias/vcs/github/wesnoth/build/../src/gui/widgets/container_base.cpp:152:21
    #11 0x72933f4 in gui2::widget::get_best_size() const /home/matthias/vcs/github/wesnoth/build/../src/gui/widgets/widget.cpp:195:12
    #12 0xa191869 in gui2::window_implementation::layout(gui2::window&, unsigned int, unsigned int) /home/matthias/vcs/github/wesnoth/build/../src/gui/widgets/window.cpp:1010:23
    #13 0xa185a9a in gui2::window::layout() /home/matthias/vcs/github/wesnoth/build/../src/gui/widgets/window.cpp:754:3
    #14 0xa175ce0 in gui2::window::draw() /home/matthias/vcs/github/wesnoth/build/../src/gui/widgets/window.cpp:577:3
    #15 0xa1bd556 in void std::__invoke_impl<void, void (gui2::window::*&)(), gui2::window*&>(std::__invoke_memfun_deref, void (gui2::window::*&)(), gui2::window*&) /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/invoke.h:73:14
    #16 0xa1bd157 in std::__invoke_result<void (gui2::window::*&)(), gui2::window*&>::type std::__invoke<void (gui2::window::*&)(), gui2::window*&>(void (gui2::window::*&)(), gui2::window*&) /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/invoke.h:95:14
    #17 0xa1bcfd4 in void std::_Bind<void (gui2::window::*(gui2::window*))()>::__call<void, gui2::widget&, gui2::event::ui_event&&, bool&, bool&, 0ul>(std::tuple<gui2::widget&, gui2::event::ui_event&&, bool&, bool&>&&, std::_Index_tuple<0ul>) /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/functional:467:11
    #18 0xa1bcc39 in void std::_Bind<void (gui2::window::*(gui2::window*))()>::operator()<gui2::widget&, gui2::event::ui_event, bool&, bool&, void>(gui2::widget&, gui2::event::ui_event&&, bool&, bool&) /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/functional:549:17
    #19 0xa1bbcc1 in std::_Function_handler<void (gui2::widget&, gui2::event::ui_event, bool&, bool&), std::_Bind<void (gui2::window::*(gui2::window*))()> >::_M_invoke(std::_Any_data const&, gui2::widget&, gui2::event::ui_event&&, bool&, bool&) /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/std_function.h:316:2
    #20 0x685071b in std::function<void (gui2::widget&, gui2::event::ui_event, bool&, bool&)>::operator()(gui2::widget&, gui2::event::ui_event, bool&, bool&) const /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/std_function.h:706:14
    #21 0x6849e9b in bool gui2::event::implementation::fire_event<std::function<void (gui2::widget&, gui2::event::ui_event, bool&, bool&)>>(gui2::event::ui_event, std::vector<std::pair<gui2::widget*, gui2::event::ui_event>, std::allocator<std::pair<gui2::widget*, gui2::event::ui_event> > >&, gui2::widget*, gui2::widget*) /home/matthias/vcs/github/wesnoth/build/../src/gui/core/event/dispatcher_private.hpp:432:4
    #22 0x67c7ae5 in bool gui2::event::fire_event<std::function<void (gui2::widget&, gui2::event::ui_event, bool&, bool&)>>(gui2::event::ui_event, gui2::event::dispatcher*, gui2::widget*) /home/matthias/vcs/github/wesnoth/build/../src/gui/core/event/dispatcher_private.hpp:504:9
    #23 0x67bfc6c in gui2::event::dispatcher::fire(gui2::event::ui_event, gui2::widget&) /home/matthias/vcs/github/wesnoth/build/../src/gui/core/event/dispatcher.cpp:113:11
    #24 0x67a1ca8 in gui2::event::sdl_event_handler::draw() /home/matthias/vcs/github/wesnoth/build/../src/gui/core/event/handler.cpp:498:15
    #25 0x911c8ed in events::raise_draw_event() /home/matthias/vcs/github/wesnoth/build/../src/events.cpp:660:13
    #26 0x911b0b5 in events::run_event_loop() /home/matthias/vcs/github/wesnoth/build/../src/events.cpp:619:2
    #27 0xa180f83 in gui2::window::show(unsigned int) /home/matthias/vcs/github/wesnoth/build/../src/gui/widgets/window.cpp:521:4
    #28 0x6a1d956 in gui2::dialogs::modal_dialog::show(unsigned int) /home/matthias/vcs/github/wesnoth/build/../src/gui/dialogs/modal_dialog.cpp:87:22
    #29 0x69ee52f in gui2::dialogs::loading_screen::display(std::function<void ()>) /home/matthias/vcs/github/wesnoth/build/../src/gui/dialogs/loading_screen.cpp:219:21
    #30 0x50721c8 in game_launcher::launch_game(game_launcher::RELOAD_GAME_DATA) /home/matthias/vcs/github/wesnoth/build/../src/game_launcher.cpp:924:2
    #31 0x4d40431 in do_gameloop(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) /home/matthias/vcs/github/wesnoth/build/../src/wesnoth.cpp:891:10
    #32 0x4d3a112 in main /home/matthias/vcs/github/wesnoth/build/../src/wesnoth.cpp:1134:19
    #33 0x7f42a09faf49 in __libc_start_main (/usr/lib/libc.so.6+0x20f49)
    #34 0x4c5e029 in _start (/home/matthias/vcs/github/wesnoth/build/wesnoth+0x4c5e029)

0x6080009a37f8 is located 88 bytes inside of 96-byte region [0x6080009a37a0,0x6080009a3800)
freed by thread T7 here:
    #0 0x4d07182 in __interceptor_free /home/matthias/LLVM/LLVM_dev/stage_2/llvm-project-20170507/compiler-rt/lib/asan/asan_malloc_linux.cc:68:3
    #1 0x7f42a73a64de  (/usr/lib/libSDL2-2.0.so.0+0x5f4de)
    #2 0x5081a51 in surface::~surface() /home/matthias/vcs/github/wesnoth/build/../src/sdl/surface.hpp:43:3
    #3 0x933cea1 in cache_item<surface>::~cache_item() /home/matthias/vcs/github/wesnoth/build/../src/image.cpp:58:8
    #4 0x933ce29 in void std::_Destroy<cache_item<surface> >(cache_item<surface>*) /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/stl_construct.h:98:19
    #5 0x933cd76 in void std::_Destroy_aux<false>::__destroy<cache_item<surface>*>(cache_item<surface>*, cache_item<surface>*) /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/stl_construct.h:108:6
    #6 0x933ccf4 in void std::_Destroy<cache_item<surface>*>(cache_item<surface>*, cache_item<surface>*) /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/stl_construct.h:136:7
    #7 0x933ca38 in void std::_Destroy<cache_item<surface>*, cache_item<surface> >(cache_item<surface>*, cache_item<surface>*, std::allocator<cache_item<surface> >&) /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/stl_construct.h:206:7
    #8 0x934f932 in std::vector<cache_item<surface>, std::allocator<cache_item<surface> > >::_M_erase_at_end(cache_item<surface>*) /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/stl_vector.h:1513:2
    #9 0x934f7ce in std::vector<cache_item<surface>, std::allocator<cache_item<surface> > >::clear() /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/stl_vector.h:1248:9
    #10 0x93343a8 in image::cache_type<surface>::flush() /home/matthias/vcs/github/wesnoth/build/../src/image.cpp:122:12
    #11 0x931266a in image::flush_cache() /home/matthias/vcs/github/wesnoth/build/../src/image.cpp:256:11
    #12 0x4fa34c9 in game_config_manager::load_game_config_for_game(game_classification const&) /home/matthias/vcs/github/wesnoth/build/../src/game_config_manager.cpp:577:2
    #13 0x507b80d in game_launcher::launch_game(game_launcher::RELOAD_GAME_DATA)::$_0::operator()() const /home/matthias/vcs/github/wesnoth/build/../src/game_launcher.cpp:930:6
    #14 0x507b1cc in std::_Function_handler<void (), game_launcher::launch_game(game_launcher::RELOAD_GAME_DATA)::$_0>::_M_invoke(std::_Any_data const&) /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/std_function.h:316:2
    #15 0x69f9a38 in std::function<void ()>::operator()() const /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/std_function.h:706:14
    #16 0x69eff94 in gui2::dialogs::loading_screen::pre_show(gui2::window&)::$_0::operator()() const /home/matthias/vcs/github/wesnoth/build/../src/gui/dialogs/loading_screen.cpp:114:5
    #17 0x69efc9c in void std::__invoke_impl<void, gui2::dialogs::loading_screen::pre_show(gui2::window&)::$_0>(std::__invoke_other, gui2::dialogs::loading_screen::pre_show(gui2::window&)::$_0&&) /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/invoke.h:60:14
    #18 0x69efa9c in std::__invoke_result<gui2::dialogs::loading_screen::pre_show(gui2::window&)::$_0>::type std::__invoke<gui2::dialogs::loading_screen::pre_show(gui2::window&)::$_0>(gui2::dialogs::loading_screen::pre_show(gui2::window&)::$_0&&) /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/invoke.h:95:14
    #19 0x69ef9ee in _ZNSt6thread8_InvokerISt5tupleIJZN4gui27dialogs14loading_screen8pre_showERNS2_6windowEE3$_0EEE9_M_invokeIJLm0EEEEDTclsr3stdE8__invokespcl10_S_declvalIXT_EEEEESt12_Index_tupleIJXspT_EEE /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/thread:234:13
    #20 0x69ef8d1 in std::thread::_Invoker<std::tuple<gui2::dialogs::loading_screen::pre_show(gui2::window&)::$_0> >::operator()() /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/thread:243:11
    #21 0x69ef3dc in std::thread::_State_impl<std::thread::_Invoker<std::tuple<gui2::dialogs::loading_screen::pre_show(gui2::window&)::$_0> > >::_M_run() /usr/lib64/gcc/x86_64-pc-linux-gnu/7.3.1/../../../../include/c++/7.3.1/thread:186:13
    #22 0x7f42a7a7db9e in execute_native_thread_routine /build/gcc/src/gcc/libstdc++-v3/src/c++11/thread.cc:83

previously allocated by thread T0 here:
    #0 0x4d076aa in __interceptor_calloc /home/matthias/LLVM/LLVM_dev/stage_2/llvm-project-20170507/compiler-rt/lib/asan/asan_malloc_linux.cc:97:3
    #1 0x7f42a73a6450  (/usr/lib/libSDL2-2.0.so.0+0x5f450)

Thread T7 created by T0 here:
    #0 0x4cefddd in __interceptor_pthread_create /home/matthias/LLVM/LLVM_dev/stage_2/llvm-project-20170507/compiler-rt/lib/asan/asan_interceptors.cc:210:3
    #1 0x7f42a7a7de95 in __gthread_create /build/gcc/src/gcc-build/x86_64-pc-linux-gnu/libstdc++-v3/include/x86_64-pc-linux-gnu/bits/gthr-default.h:662
    #2 0x7f42a7a7de95 in std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) /build/gcc/src/gcc/libstdc++-v3/src/c++11/thread.cc:163
    #3 0x69ea668 in gui2::dialogs::loading_screen::pre_show(gui2::window&) /home/matthias/vcs/github/wesnoth/build/../src/gui/dialogs/loading_screen.cpp:110:21
    #4 0x6a1d694 in gui2::dialogs::modal_dialog::show(unsigned int) /home/matthias/vcs/github/wesnoth/build/../src/gui/dialogs/modal_dialog.cpp:83:2
    #5 0x69ee52f in gui2::dialogs::loading_screen::display(std::function<void ()>) /home/matthias/vcs/github/wesnoth/build/../src/gui/dialogs/loading_screen.cpp:219:21
    #6 0x50721c8 in game_launcher::launch_game(game_launcher::RELOAD_GAME_DATA) /home/matthias/vcs/github/wesnoth/build/../src/game_launcher.cpp:924:2
    #7 0x4d40431 in do_gameloop(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) /home/matthias/vcs/github/wesnoth/build/../src/wesnoth.cpp:891:10
    #8 0x4d3a112 in main /home/matthias/vcs/github/wesnoth/build/../src/wesnoth.cpp:1134:19
    #9 0x7f42a09faf49 in __libc_start_main (/usr/lib/libc.so.6+0x20f49)

SUMMARY: AddressSanitizer: heap-use-after-free /home/matthias/vcs/github/wesnoth/build/../src/sdl/surface.hpp:85:4 in surface::add_surface_ref(SDL_Surface*)
Shadow bytes around the buggy address:
  0x0c108012c6a0: fa fa fa fa 00 00 00 00 00 00 00 00 00 00 01 fa
  0x0c108012c6b0: fa fa fa fa 00 00 00 00 00 00 00 00 00 00 05 fa
  0x0c108012c6c0: fa fa fa fa 00 00 00 00 00 00 00 00 00 00 01 fa
  0x0c108012c6d0: fa fa fa fa 00 00 00 00 00 00 00 00 00 00 05 fa
  0x0c108012c6e0: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c108012c6f0: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd[fd]
  0x0c108012c700: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c108012c710: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c108012c720: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c108012c730: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c108012c740: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==17470==ABORTING
````

**Comment by Vultraz** [comment]
This is on master?

**Comment by jyrkive** [comment]
The log says that **thread 7** freed the surface (by flushing the image cache) but **thread 0** accessed it. This may be some kind of race condition where the image cache is simultaneously being flushed and read.

**Comment by matthiaskrgr** [comment]
Yeah, it's on master.

**Comment by matthiaskrgr** [comment]
Yeah, it's on master.

**Comment by jyrkive** [comment]
(Deleted a duplicate comment)

**Comment by matthiaskrgr** [comment]
Ok, I managed to reproduce on the 1.14 branch as well:
Simply start the tutorial,
quit to main menu
restart the tutorial:
=> crash

**Comment by Vultraz** [comment]
Could you see if this is still reproducible on master (after 6d0b7c8) or 1.14 (after 0eb27e2)?

**Comment by matthiaskrgr** [comment]
Fix confirmed on 1.14 (can't test master since the ui is still broken and I can't even exit the tutorial)
